Set Dock icon on the macosx backend

joelfrederico opened the pull request:
## PR Summary

This issue is similar to #14850/#14927. The issue is not isolated to the macos backend. It also happens in Qt5 (at least, possibly others).

Turns out Qt5 broke setWindowIcon. It seems like it is similar to this bug report: https://bugreports.qt.io/browse/QTBUG-55932. It fails silently, you request an icon change and nothing happens on macOS.

I have doubts this will be fixed without some pretty large Qt5 changes. Apple's API doesn't support .svg files. It does support a wide variety of other graphics, including vector graphics: pdf, and postscript, for instance. But Qt's API only supports .svg vector graphics.

We can work around this with png's, etc., but we already have a far superior .pdf logo. Apple users have come to expect high-quality vector graphics on their Retina displays. We should use the vector .pdf logo.

We can work around by using the ctypes lib to call Apple routines directly to avoid adding any more dependencies. It's done fairly lazily, so as not to be a performance hit on non-Apple platforms.

This appears to work for me. Could use comments and/or tests, although I have no idea how to write a unit test for an app icon in macOS.

## PR Checklist

- [x] Has Pytest style unit tests
- [x] Code is [Flake 8](http://flake8.pycqa.org/en/latest/) compliant

joelfrederico commented:
I'm not sure about unit tests. Could somebody provide feedback? I don't really have a way to unit-test any of the code, aside from perhaps checking that the ctypes/Apple api's say that I did, in fact, successfully change the icon. It sort of seems like circular reasoning: "Do the API's say they worked?"

At any rate, the icon is set for every plot on macOS with Qt5, so in a sense we are piggy-backing on those unit tests too.

Thoughts?

joelfrederico commented:
I didn't see a reason to reset the icon, although that is also possible. Thoughts?

tacaswell commented:
My knee-jerk reaction as that this is too much complexity for the problem it is solving.

From the linked bug report it suggests that the issue is that the at-conversion of svg -> QIcon does not provide a list of available sizes and then there is a bug in the OSX specific code that results in no icon being shown.  Could we create a QIcon and set the missing size information?

Failing that, rendering a hi-dpi png (which we may already have due to the other backends) and using that seems like a better option.

joelfrederico commented:
I did play around with trying to get the QIcon to render/rasterize the SVG and failed. I can make another pass at it.

I did check for a high-res png and came up empty. None of them were large enough on my display. I wouldn't be surprised if the default on other backends were to use SVG, that's what Qt5 uses. I would recommend generating a larger logo if we go this route. I'm not sure how much larger- I'd want to do some math for render sizes when using a 5k monitor. Maybe just follow Apple's recommended icon sizes.

Neither are perfectly ideal if the system supports vector icons and we're providing a raster. From a performance standpoint, I'd say creating the logo is worst, rastering is a better, and loading a high-res png is best.

I can also put this in the macos backend code instead of ctypes. If we need the code for the macos backend anyways, then I could expose the Objective-C functionality as an internal Python API for other backends to use. Of course this also violates encapsulation of the backends, which may or may not be a big deal here, and might require a Python wrapper so that other platforms don't complain about the missing macos module.

I can implement everything except for generating a new icon. I could probably figure that out, I'm just not sure of the process for doing that and adding a photo to the repo.

What would you like me to do?

tacaswell commented:
> I can also put this in the macos backend code instead of ctypes. If we need the code for the macos backend anyways, then I could expose the Objective-C functionality as an internal Python API for other backends to use. Of course this also violates encapsulation of the backends, which may or may not be a big deal here, and might require a Python wrapper so that other platforms don't complain about the missing macos module.

This seems like a promising path to explore.

generating (and caching) the png on the fly seems like a good option too.

anntzer commented:
The icons are generated with tools/make_icons.py (https://github.com/matplotlib/matplotlib/pull/14941 may help a bit).  You'll need to edit it to generate bigger sized icons.

joelfrederico commented:
> > I can also put this in the macos backend code instead of ctypes. If we need the code for the macos backend anyways, then I could expose the Objective-C functionality as an internal Python API for other backends to use. Of course this also violates encapsulation of the backends, which may or may not be a big deal here, and might require a Python wrapper so that other platforms don't complain about the missing macos module.
> 
> This seems like a promising path to explore.

My last push went with this solution. If you like it, I'll add tests.

joelfrederico commented:
Wrote tests, fixed a few things. This only makes the matplotlib icon show up on macosx and Qt5Agg backgrounds.

anntzer commented on lib/matplotlib/backends/backend_macos_common.py:
```diff
@@ -0,0 +1,10 @@
+import sys
```
this should definitely be a private module (with a leading underscore)

anntzer replied:
rename the file to e.g. _macos_helpers.py or whatnot.

joelfrederico replied:
Gotcha, will do.

joelfrederico replied:
See latest push.

anntzer commented on lib/matplotlib/tests/test_backend_macos_common.py:
```diff
@@ -0,0 +1,37 @@
+        with pytest.raises(TypeError):
+            bmc.set_mac_icon()
+
+        with pytest.raises(RuntimeError):
```
Not sure why these don't raise TypeError (nor whether we really need to test them).

joelfrederico replied:
So the thing is, `matplotlib.cbook._get_data_path` returns a `PosixPath` object. It’s natural for people to pass in something that represents the path but isn’t a string. I can convert whatever argument is given me into a string, but that converts everything into a string, including these types. I could check its type to avoid that, but that’s not very Pythonic. Sort of a catch-22. Not sure of a good solution...

anntzer replied:
My point is that if someone passes you something of the wrong type, it makes sense to raise a TypeError, not a RuntimeError.

joelfrederico replied:
Yeah, absolutely, I want to do that. I just can’t think of a way to handle arbitrary types like `PosixPath` and also provide a TypeError. There’s nothing I know of that will only convert the various path types in Python while also erroring on things that aren’t paths.

Any ideas? Maybe I should give up trying to handle arbitrary types and only allow strings?

anntzer replied:
see https://github.com/matplotlib/matplotlib/pull/14930#discussion_r309894455

anntzer commented on src/_macosx.m:
```diff
@@ -178,6 +178,101 @@ static int wait_for_stdin(void)
+    }
+
+    const char* path_string;
+    PyObject* unicode_path = PyObject_Str(path);
```
personally I prefer a style like https://github.com/matplotlib/matplotlib/blob/master/src/_tkagg.cpp#L306 (more code, less blank lines...) though it's just my opinion.

anntzer replied:
Also, probably this should use https://docs.python.org/3/c-api/unicode.html#c.PyUnicode_DecodeFSDefault instead of PyObject_Str?

joelfrederico replied:
I can’t use that one- the input arguments are different, `PyObject_Str` takes a `PyObject`, but `PyUnicode_DecodeFSDefault` takes a `const char*`.

anntzer replied:
But you still need to perform proper decoding...
I guess with any of these, don't try to retrieve the icon path in C code at all; just expose a function that takes a bytes as argument (the decoded path) and sets the icon from it, and call that function from the python-level backend_macos.py.

joelfrederico replied:
I think that's worse, the bytes could be anything and might not decode as proper UTF8. The NSImage API doesn't document what happens if I pass in invalid UTF8...

Working backwards:

- I'd like to keep the Python decoding next to the NSImage API. Python `PyUnicode_AsUTF8 ` API says it'll give me a UTF-8 and NSImage needs a UTF-8.
- The Python `PyUnicode_AsUTF8` takes PyObject*, so I'd like to give it a `str`. It'll fail on anything else, including things like `PosixPath`.

So I think I have two choices:

- C function takes any objects and converts them to `str` somehow
- C function only works if you give it a `str`, fails on anything else

I wish I could do the first without converting everything or doing type-checking. E.g., `str(None)` becomes "None".

How would you do this?

anntzer replied:
Then make your function take a str as argument and decode it using the `s` code per https://docs.python.org/3/c-api/arg.html#strings-and-buffers (which will decode with utf8 into a const char*).

joelfrederico replied:
Ooooo this note on `s` is helpful:

> Note This format does not accept bytes-like objects. If you want to accept filesystem paths and convert them to C character strings, it is preferable to use the O& format with PyUnicode_FSConverter() as converter.

That's exactly what I want to do. I'll fix it using Python's recommendation, yeah?

anntzer replied:
yes

joelfrederico replied:
See latest push.

anntzer commented on src/_macosx.m:
```diff
@@ -178,6 +178,101 @@ static int wait_for_stdin(void)
+    Py_DECREF(unicode_path);
+    if (!path_string)
+    {
+        PyErr_SetString(PyExc_TypeError, "bad argument type, must be "
```
you probably don't need to set the exception, PyUnicode_AsUTF8 will already have done it for you

joelfrederico replied:
I set this one because the exception they use is opaque. IIRC, it’s something like, “bad argument type in compiled code”. I was iffy on setting my own? I could go either way, let me know.

joelfrederico replied:
See latest push.

anntzer commented on src/_macosx.m:
```diff
@@ -178,6 +178,101 @@ static int wait_for_stdin(void)
+
+    // API indicates Python owns path_string; we don't need to free it
+    path_string = PyUnicode_AsUTF8(unicode_path);
+    Py_DECREF(unicode_path);
```
that looks like path_string may be dangling? (because unicode_path will be gc'd and so path_string will be free'd)

joelfrederico replied:
Ha. Good call. Weird that this worked anyways.

joelfrederico replied:
See latest push.

anntzer commented on src/_macosx.m:
```diff
@@ -178,6 +178,101 @@ static int wait_for_stdin(void)
+    }
+    @catch (NSException *e)
+    {
+
```
please cut down a bit on the blank lines :p

joelfrederico replied:
Sure, I’ll condense it.

joelfrederico replied:
See latest push.

joelfrederico commented:
@anntzer Okay, pushed fixes, could use another review.

anntzer commented on lib/matplotlib/backends/_macos_helpers.py:
```diff
@@ -0,0 +1,10 @@
+import sys
```
I would not even bother with testing the os and just make the module only importable on OSX (just like backend_macos.py).

joelfrederico replied:
Also, I have to wrap at some level because this will wind up in other backend code that’s platform-agnostic, like the Qt5 part. I’m being extra-paranoid, though, I actually check twice. Maybe only one level of paranoia?

How should we resolve this?

anntzer replied:
Just wrap in once before importing it and call it a day.

joelfrederico replied:
I could get rid of the _macos_helpers.py entirely if we just make sure to wrap imports of _macosx with platform checks, for instance... It's sort of a style and paranoia level choice. Lots of ways to skin the cat, just not sure how you'd like it done.

anntzer replied:
I personally don't like duplicated checks, but won't insist on having absolutely nothing done twice.

joelfrederico replied:
Kind of like this?

```
import sys

if sys.platform == "darwin":
    from ._macos_helpers import set_icon as set_mac_icon
else:
    def set_mac_icon(path):
        pass
```

Then in the Qt5 backend, I can do:

```
from matplotlib.backends import _macos_helpers

...

_macos_helpers.set_mac_icon(...)
```

anntzer replied:
I would not have _macos_helpers at all (one less module, yay), and in qt5 I would have
```
if sys.platform == "darwin":
    import _macos; _macos.set_mac_icon(...)
```

joelfrederico replied:
I'm good with that. Where would I put the tests? There don't seem to be tests for macOS. Or do I still need them?

anntzer replied:
I wouldn't bother with these tests.  I mean, they're not even actually testing the real functionality, but just that the correct error is thrown when a private function is called with stuff that it's never going to be called with...

joelfrederico replied:
Well, mostly I just wanted to make sure the code wouldn’t crash when given bad input... I could imagine somebody moves the matplotlib.pdf in the repo, and then all of their tests crash hard...

On the other hand, if the code did crash, these tests wouldn’t really help with narrowing down the issue...

Your call. Test or don’t test?

anntzer replied:
Sure, you can put one, I guess in test__macosx.

joelfrederico replied:
Okay, done. Latest push has the test. Another review?

anntzer commented on lib/matplotlib/tests/test__macos_helpers.py:
```diff
@@ -0,0 +1,37 @@
+        pass
+
+
+if sys.platform == "darwin":
```
and likewise not bother here

anntzer commented on src/_macosx.m:
```diff
@@ -737,7 +825,10 @@ static CGFloat _get_device_scale(CGContextRef cr)
     rect.size.height = height;
     rect.size.width = width;
-
+    if (_set_icon() != 0)
```
Why don't you do this in the python-side `FigureManagerMac.__init__` (just like for qt)? then you can get rid of the c-level _get_pdf_icon_path.

joelfrederico replied:
Good idea. I didn’t know that was an option. Should clean things up a bit.

joelfrederico commented:
Okay @anntzer, another push another review?

anntzer commented:
looks fine to me, but now you need to find an actual mac user (ie not me) to test this :p

joelfrederico commented:
Any suggestions? I assume I don’t count :P

anntzer commented:
Tagged the PR, I'm sure they'll show up :p

joelfrederico commented:
@anntzer Should we bump this somehow? Seems like nobody has noticed, but I'm not sure how long that usually takes here.

jklymak commented:
I've been away, but wasn't popping this to the top of my queue because I wasn't clear that Matplotlib is an application, and hence am not clear why we would set the dock icon.

joelfrederico commented:
@jklymak Honestly didn't consider it, since the Qt backends set the app icon on other platforms. Maybe the icon shouldn't be set in general?

joelfrederico commented:
But to be fair, they set the window icon and not the taskbar icon for e.g. Windows for me. I assumed this was a bug, because the icon is the Windows default blank page icon, which just seems like it was forgotten.

While macOS doesn't actually have a window icon...

I don't know!

jklymak commented:
Yeah, I'm not saying this PR is wrong, just not sure what it should do an a Mac, or if its worth the maintenance headache if this goes awry down the line

tacaswell commented on lib/matplotlib/tests/test__macosx.py:
```diff
@@ -0,0 +1,34 @@
+        pass
+
+
+if sys.platform == "darwin":
```
can you use `pytest.skipif` instead?

joelfrederico replied:
Sure! Might be a bit before I can make a commit.

joelfrederico commented:
@jklymak Apple has these guidelines (https://developer.apple.com/design/human-interface-guidelines/macos/icons-and-images/app-icon/):

App Icon
Beautiful, compelling icons are a fundamental part of the macOS user experience. Far from being merely decorative, icons play an essential role in communicating with users. To look at home in macOS, an app icon should be meticulously designed, informative, and aesthetically pleasing. It should convey the main purpose of the app and hint at the user experience.

App icons are associated with windows, which mean they should represent what the window's main purpose is. I think this is a solid argument for changing the app icon.

As a first pass, I think this PR is good. There is a slight complication where matplotlib may be embedded in another application. This is only relevant in some backends (IIRC the macosx backend cannot be embedded). In that case, the app icon should reflect the embedding application. That means we either need to detect embedding, or provide an API for using another app. This is probably a backend-specific consideration.

I don't think we should avoid improvements because of maintenance concerns... If it's a valid thing to do on macOS, if it makes sense and improves UX, it should be done right in order to minimize maintenance.

tacaswell commented:
I could have sworn I posted a comment earlier today but can not find it in this thread now (but it looks like @joelfrederico is responding to it!).

> App icons are associated with windows

How does in know which window's icon to set?  From the code it looks like this applies at the application level?  If it is one per-application, can we do it just once in the module?

> (IIRC the macosx backend cannot be embedded)

I don't actually know objective-c, but from skimming `_macosx.m` it looks like it follow the same pattern (a 'widget' for the canvas which is then embedded in a figure along with a toolbar).  It may require writing a `.m` file, but I suspect it is possible to embed a matplotlib figure into a large native OSX application, but that is neither here nor there for this discussion.


Do tk, gtk, and wx have the same issue of lacking an icon?

joelfrederico commented:
> 
> > App icons are associated with windows
> 
> How does in know which window's icon to set? From the code it looks like this applies at the application level? If it is one per-application, can we do it just once in the module?

Yes, it’s one icon per process. So we could do this just once, maybe even at the backend import level. I haven’t gotten too far into it, but right now matplotlib only has a Dock icon when a plot is visible. I don’t know how that works or why it was designed that way...

> 
> > (IIRC the macosx backend cannot be embedded)
> 
> I don't actually know objective-c, but from skimming `_macosx.m` it looks like it follow the same pattern (a 'widget' for the canvas which is then embedded in a figure along with a toolbar). It may require writing a `.m` file, but I suspect it is possible to embed a matplotlib figure into a large native OSX application, but that is neither here nor there for this discussion.

I think this is only possible if you also embed python. But yes, now that I think about it, you could embed python and create matplotlib windows, which would then override the process’s icon. You might also be able to embed the widget... it depends on the matplotlib api, the current macosx backend may or may not have an api that makes this possible.

> Do tk, gtk, and wx have the same issue of lacking an icon?

Ah I don’t know. I should probably make a more serious attempt at testing the various backends. Probably should also cache the original icon and reset when finished, whatever “finished” means.

tacaswell commented:
> I haven’t gotten too far into it, but right now matplotlib only has a Dock icon when a plot is visible. I don’t know how that works or why it was designed that way...

I suspect the doc-icon is generated when the QApplication (or the OSX backend equivalent) is instantiated.  We are running a bit inverted, normally the _first_ thing your process in a 'classic' desktop app is to instantiate the main "Application" instance (which in Qt you need to do before you can instantiate any of the widgets), however the user is usually running python first, and then at some point later when the user actually asks us for a window we fire up the qapp (and install the input hooks to make it look 'live' running inside your terminal) see https://github.com/matplotlib/matplotlib/blob/master/lib/matplotlib/backends/backend_qt5.py#L98 and https://github.com/matplotlib/matplotlib/blob/master/lib/matplotlib/backends/backend_qt5.py#L229  I suspect the easiest way to make sure we are not going to stomp on a user-set icon is to only try to set it if _we_ create the qApp.

joelfrederico commented:
> I suspect the doc-icon is generated when the QApplication (or the OSX backend equivalent) is instantiated.

This is true for Qt. It's also possible to tell if Q(Gui)Application has been instantiated, at least in the PyQt5 backend. `PyQt5.QtCore.QApplication.instance() == None` if it has not been instantiated.

This is not true for macOS's NSApplication while running in Python. There is an API for switching application "activation policy" to a GUI mode. The Dock icon is generated at that point. It's also possible to tell if it's already in an activated mode via an adjacent API.

I think the right thing to do is to cache and set the icon at the point where the application instance (QApplication/NSApp) is instantiated and/or accessed, if there is an icon to access.

Then, there are various API's for running code once we've closed all of our windows (both for macosx and PyQt5). That code can access the cached icon (if any) and restore it.

How does that solution sound?

I'm not sure about other backends. I'd like to defer those until the backends I'm familiar with, macOS and PyQt5, are solid.

joelfrederico commented:
FYI, from what I've seen, all of the backends I can test (Qt5Agg, macosx, TkAgg, WxAgg) currently have the same behavior (with the exception of WebAgg and nbAgg):

- Backend uses Python icon
- Backend does not turn off Python icon when all windows are closed
  - Unclear if this is because backends don't stop running when windows are closed
- Icon is the Python rocketship

WebAgg does not show an icon at all. nbAgg is specifically for embedding in Jupyter notebook/lab, so showing an icon doesn't make sense.

Digression: FWIW, it looks like Qt4 is unsupported and a variety of package managers have dropped support for it. Perhaps matplotlib should drop Qt4 as a backend?

joelfrederico commented:
>   * Edit: I know why: [setActivationPolicy](https://developer.apple.com/documentation/appkit/nsapplication/1428621-setactivationpolicy?language=objc) can only show icons and enable activation, not remove them or disable activation

Okay, so that's actually not true. You can change the policy for macOS 10.9+: https://github.com/lionheart/openradar-mirror/issues/16080.

joelfrederico commented:
FYI: I’ve discovered that the run loop integration isn’t handled very well, and there is quite a bit of cruft in the macOS Objective-C code. I have a consistent method for fixing all of these things, and integrating icon manipulation. In order for that all to make sense, though, I think there are a few cleanup tasks I’d like to get done. If I can tighten up the existing system, integrating icon manipulation will be a much more logical extension of the rest of the code.

This PR shouldn’t be merged yet, but I will rebase onto the changes that get accepted.

How does that sound?

greglucas commented:
@joelfrederico, any interest in picking this back up? I just checked out the branch and it looks like a nice addition to add the icon to the app tray!

joelfrederico commented:
@greglucas Actually, that would be a nice project over Thanksgiving. But I don’t think I have buy-in from @tacaswell. @tacaswell: Could you make a post about the behavior you’d like, so I can code it up?

greglucas commented:
Going back through the conversation a bit, it seems like the scope has perhaps expanded a bit as you've been going through this. My suggestion would be to try and descope this a bit and turn it into multiple PRs. Focus on adding this basic capability to the generic app first, which you already have. Then I would make a follow-up PR if you want to handle the embedding case. And other PRs if you want to clean things up. There has been quite a bit of cleanup in the backend codebase since you started working on it, so the rebase may be a bit tricky (You may want to start on a new fresh branch and copy-paste your code into the right spots rather than spending time resolving conflicts).

I think there is good motivation for having this, as the branding is nice. Tkagg does apparently add the icon, but an older or lower resolution version of it.
https://github.com/matplotlib/matplotlib/blob/c19cca75e54fc4df21559c656261242e0029ca0d/lib/matplotlib/backends/_backend_tk.py#L927-L938
Original:
![image](https://user-images.githubusercontent.com/12417828/143050938-19b3efba-7562-4c33-aa64-bc0f38f253bf.png)
This PR:
![image](https://user-images.githubusercontent.com/12417828/143050465-9c89861d-d159-46ce-a824-e2398eb46617.png)
tkagg:
![image](https://user-images.githubusercontent.com/12417828/143051613-531ef9df-3cca-464d-98e0-aacce53ff09c.png)

joelfrederico commented:
@greglucas Yes, things are probably quite different now. I was thinking this probably got derailed due to scope creep. Given the precedent of tkagg, I could probably just copy their behavior. The most difficult question is when to change the icon, and when not to.

tacaswell commented:
My main concern is that we (as a library) do not conflict with users who are trying to make their own application.  Declaring "pyplot _is_ an application" is probably an acceptable escape hatch if needed, but it would need to be well justified.

------

I'm going to delegate decision making on this to @greglucas ;)

joelfrederico commented:
@greglucas I think there's a fundamental difficulty with the icon work. I don't think it's universally possible to detect whether Matplotlib is running on its own or in concert with some other logic. We could have default behavior, but I don't think it's possible to satisfy everybody.

For that matter, on my system in some scenarios, Matplotlib will create a bouncing Python rocketship until `show()` is called. There's lots of anecdotes like that (see above), that sort of defy an easy or clear expectation for what should happen.

What behavior would you like? I'd be happy to do whatever, if there's just a clear goal for what I should try to accomplish.

jklymak commented:
The confusion here is whether we run the GUI or someone else is.  This is a pretty good reason to disambiguate pyplot, top-level, and the gui.  

Is there really any reason for a downstream library embedding Matplotlib to call pyplot?  Can that be the heuristic of whether to show the icon or not?  If pyplot is spun up, then go ahead and show the icon; if a downstream library gets mad that their icon is squashed, ask them to not use pyplot.

greglucas commented:
Here is my (maybe overly simplistic) take on this for each of the cases. Basically, push the logic up into the manager instead of the canvas portion of the code. If we are managing things, I think that we can set the icon if we want.

### QT backend on mac
It looks like the icon needs to be set on the main App and not just the windows as is done with the other platforms... So, I think this could work, as this would only put it on any new QT Applications() we start up. If someone else is embedding the canvas I don't think this will have any effect because we will have bailed out earlier when an app was already detected.

```diff
diff --git a/lib/matplotlib/backends/backend_qt.py b/lib/matplotlib/backends/backend_qt.py
index 3f37792983..4c7c9a5a6a 100644
--- a/lib/matplotlib/backends/backend_qt.py
+++ b/lib/matplotlib/backends/backend_qt.py
@@ -125,6 +125,8 @@ def _create_qApp():
             cbook._setup_new_guiapp()
         else:
             qApp = app
+        image = str(cbook._get_data_path('images/matplotlib.svg'))
+        qApp.setWindowIcon(QtGui.QIcon(image))

     try:
         qApp.setAttribute(QtCore.Qt.AA_UseHighDpiPixmaps)  # Only for Qt<6.
```

I think this is also preferable to the `_macosx.set_icon()` call from QT because I think that may actually start up the macosx event loop? But, I haven't checked that.

### macosx backend

I think similar logic could be applied here. Instead of putting the `set_icon` call in the canvas portion, push it up into the `FigureManagerMac` class. Then if someone is embedding, they will only be using the `FigureCanvasMac` and avoid the icon I think.

joelfrederico commented:
Sounds good to me. The original issue seemed to be related to setWindowIcon being a noop on macOS when I originally started this. I'll check to see if it's still an issue, perhaps it was fixed.

joelfrederico commented:
I'm sure it needs cleanup and style changes. I'd welcome any suggestions/tweaks! There are quite a few choices that could be done many ways and I don't know what would be preferred.

greglucas commented on lib/matplotlib/backends/backend_macosx.py:
```diff
@@ -66,7 +66,10 @@ class FigureManagerMac(_macosx.FigureManager, FigureManagerBase):
     def __init__(self, canvas, num):
-        _macosx.FigureManager.__init__(self, canvas)
+        icon_path = None
+        if 'com.adobe.pdf' in _macosx.FigureManager._get_image_types():
+            icon_path = str(cbook._get_data_path('images/matplotlib.pdf'))
+        _macosx.FigureManager.__init__(self, canvas, icon_path)
```
My personal preference would be to keep it as you had it before with an explicit `_set_icon()` function rather than changing the init signature. But, I'm not sure I have a huge preference, others may have stronger feelings around this.

joelfrederico replied:
I don't have strong feelings, I'll do whatever the consensus is. (Or just use your suggestion if I don't hear anything within a day or two.)

greglucas commented on src/_macosx.m:
```diff
@@ -1,3 +1,4 @@
+#include <objc/NSObjCRuntime.h>
```
What functions are we grabbing from the runtime header? Everything below looks pretty standard I thought.

joelfrederico replied:
That's weird, I don't remember adding this... going to revert.

greglucas commented on src/_macosx.m:
```diff
@@ -629,6 +630,39 @@ static CGFloat _get_device_scale(CGContextRef cr)
+        return -1;
+    }
+    NSImage* image = [[NSImage alloc] initByReferencingFile: ns_icon_path];
+    [ns_icon_path release];
```
I thought you were only supposed to release explicitly alloc'd items?

joelfrederico replied:
Maybe this is the source of your segfault? I'm not super familiar with Obj-C semantics, but I think you're right at any rate.

greglucas commented on src/_macosx.m:
```diff
@@ -629,6 +630,39 @@ static CGFloat _get_device_scale(CGContextRef cr)
+        PyErr_SetString(PyExc_RuntimeError, "Could not convert to NSString*");
+        return -1;
+    }
+    NSImage* image = [[NSImage alloc] initByReferencingFile: ns_icon_path];
```
Can you add `[... autorelease]` here and let the program do the releasing for you?

joelfrederico replied:
I'd have to add an @autorelease {} scope block, right?

greglucas replied:
I am not an expert in objc either... But, I _think_ your new update is the way to do it and the scope will be accounted for properly. The autorelease pools are optional if you want finer grained control I think.

greglucas commented on src/_macosx.m:
```diff
@@ -778,6 +820,23 @@ static CGFloat _get_device_scale(CGContextRef cr)
 }
 
+static PyObject*
+FigureManager__get_image_types(PyObject* no_self, PyObject* no_args)
```
Do we need this helper function? It looks like above you are only trying to set the pdf icon, so we could even do a simple check for that case explicitly within the `set_icon()` function and return 0 if it isn't supported and just move on.

joelfrederico replied:
According to the docs, the compatible image formats vary with OS release. I'm under the impression that in practice they're fairly stable. On the other hand, if we know the available formats, we can pick the right image from the install based on what will work. That would beat just trying a bunch of formats and hoping they don't error?

greglucas replied:
But you aren't cycling through the others above? It looks like it is only trying the pdf version right now.

joelfrederico replied:
I mean, maybe I’m anticipating too much? It’s more pythonic to try and fail than make it complicated. So rather, don’t check if it’s a compatible file type but just set it. If it’s not a valid file, it’ll fail. If we need to support different file types on different versions of macOS (for whatever reason), add the complexity then. Yes?

greglucas replied:
Yep, exactly :) I think that will simplify this PR, and we can add more PRs in the future to expand on capabilities if we find out it is needed.

joelfrederico replied:
Done

greglucas reviewed (commented):
I do get an icon showing up, which looks nice!

But, when interacting there is something else going on when I open new windows. I get a segfault when clicking on the subplot tool icon with this update, so something might not be working quite right between the manager and the new window that gets popped up with subplot tools.

joelfrederico commented:
I think the segfault you were experiencing was due to other code which has since been resolved. I rebased to a new point in main, could you test again?

Could we also modify the PR to not target Qt? I've moved Qt into another PR.

greglucas commented:
Yep, it seems to work well now without a segfault. Lets see if others have additional comments over the next few days about the architecture comments, but I think this is looking good and splitting it into separate PRs is also nice, thanks for doing that!

greglucas commented on src/_macosx.m:
```diff
@@ -637,24 +668,32 @@ static CGFloat _get_device_scale(CGContextRef cr)
+                &PyUnicode_FSDecoder, &icon_path)) {
+        return -1;
+    }
+    if (set_icon(icon_path)) {
+        return -1;
+    }
```
Won't this cause us to fail if the icon_path wasn't found or wasn't valid? I think that setting the icon should be more of a "nice-to-have", than a "must-have" feature. I think this is partially why I was thinking the `set_icon` call would be better up in the manager as an "extra" rather than in the init itself.

joelfrederico replied:
Yes, it is. As for whether it's optional or not, the icon is a keyword argument with default None. By default, the icon is not set.

I think there are two separate issues: what happens on an error while setting an icon, and how/where an icon should be set.

For the first, I think an error in the Obj-C should result in a Python exception. I don't think setting an icon should silently fail. It would be impossible to debug without compiling against a debug Python, with debug dependencies, and running in a debugger. That's a hassle. If it makes sense to handle the error at a higher level, that can be done.

For the second, I am a bit more ambivalent. The real question is, "Should the icon be handled via a context manager?" In other words, should it be reset every time after Matplotlib finishes running? If the answer is no, then it doesn't much matter to me: make it a property, make it a setter method, put it in the constructor, it's just style semantics in my opinion and how you want to handle (or not) the Python exception raised. If the answer is yes, though, then this design is wrong and should change. That would be an API change, which would be a whole different can of worms...

greglucas replied:
Ahh, yes, I missed that the default was None, thanks for clarifying! I agree with everything you said :) 1. Raise an error for sure, and 2. set it in the initializer like you have, or a direct call to set_icon() from Python right after initialization.

The reason I have a preference for the set_icon() coming from Python is to maintain similarity to the other backends where they call `setWindowIcon()` or `iconphoto()` after the window setup, and the manager init I think remains consistent then. But, as you mentioned we should see if others have a strong preference. I think we are down to the details at this point and overall it is on the right path 🚀

joelfrederico commented:
@greglucas I figured it's been long enough, so I went back to a set_icon method. I made it static because it doesn't reference any saved state.

joelfrederico commented on src/_macosx.m:
```diff
@@ -611,24 +611,25 @@ static CGFloat _get_device_scale(CGContextRef cr)
         return -1;
     }
+    static char* kwlist[] = { "canvas", NULL };
+    if (!PyArg_ParseTupleAndKeywords(args, kwds, "O!", kwlist,
```
Technically this should be in a separate PR?

greglucas replied:
I'm not sure I follow, why did you update this here? Now that you changed back to a method, shouldn't this just be left alone? So, I guess yes, if you want to update the init method it should be a separate PR.

joelfrederico replied:
There was an acceptance of any Python object with an unsafe cast to a _macos.FigureCanvas. I at least added a type check which should be safe in most scenarios.

I'll put it in a separate PR.

joelfrederico commented on src/_macosx.m:
```diff
@@ -697,6 +698,41 @@ static CGFloat _get_device_scale(CGContextRef cr)
+            return NULL;
+        }
+        NSApplication* app = [NSApplication sharedApplication];
+        app.applicationIconImage = image;
```
Is there any way to catch errors that might happen here?

greglucas replied:
I suppose you could put this in a `@try` block and then catch the objc exception and raise a Python exception inside the except clause.
https://developer.apple.com/library/archive/documentation/Cocoa/Conceptual/Exceptions/Tasks/HandlingExceptions.html

greglucas commented on src/_macosx.m:
```diff
@@ -697,6 +698,41 @@ static CGFloat _get_device_scale(CGContextRef cr)
+FigureManager_set_icon(PyObject* null, PyObject* args, PyObject* kwds) {
+    PyObject* icon_path;
+    static char* kwlist[3] = { "icon_path", NULL };
+    if (!PyArg_ParseTupleAndKeywords(args, kwds, "O&", kwlist,
```
We should change to just `PyArg_ParseTuple` here if we aren't going to use the keywords.

joelfrederico commented:
@greglucas Should be good now?

greglucas reviewed (approved):
Yep, this looks good to me and works well with multiple windows and subplot params!

I thought I made an earlier comment, but now I can't seem to find it, so I must not have hit enter... I was wondering if you think we should put a try/except clause around the Python `set_icon()` call? Right now if I put a bad file-path in (lets say someone removed the MPL data path), I get a Runtime Exception and MPL stops. QT5agg continues on without issue and tkagg logs the exception and continues on. I think that we should perhaps adopt the tkagg philosophy here and log this and continue on...
https://github.com/matplotlib/matplotlib/blob/40c77d34f3d8e6b970dc1e8ebf6fd37d5eef368e/lib/matplotlib/backends/_backend_tk.py#L934-L938

joelfrederico commented:
Right now passing tests let us know that set_icon is working. If you want to do that we'd have to add a test for set_icon because it would no longer be covered.

joelfrederico commented:
@greglucas What do I need to do to wrap this up?

greglucas commented:
You've got my approval, but someone else will need to test it out before merging still, so just waiting on the second review which can take a bit of time since it is all volunteer basis. Thanks for your patience!

joelfrederico commented:
@tacaswell @anntzer Any chance you’d know somebody who we could ping to merge this?

anntzer commented:
Perhaps @dstansby or @jklymak can help on that?

dstansby reviewed (approved):
This works for me. Thanks a lot for this PR - after so many years using Matplotlib without a dock icon, it feels a bit surreal to finally see one!

tacaswell commented:
Thanks @joelfrederico for sticking with us and getting this done!
